File: 834. Sum of Distances in Tree.py
class Solution:
    def sumOfDistancesInTree(self, n: int, edges: List[List[int]]) -> List[int]:
        def count_vertices_subtree(curr, previous):
            nonlocal graph, count
            c = 1
            for next in graph[curr]:
                if next == previous:
                    cotinue
                c += count_vertices_subtree(next, curr)
            count[curr] = c
            return c
        def rootDistance(curr, previous, distance):
            nonlocal graph, result
            result[0] += distance
            for next in graph[curr]:
                if next == previous:
                    continue
            rootDistance(next, curr, distance+1)
        def totalDistance(curr, previous):
            nonlocal graph, result, count, n
            if previous != -1:
                result[curr] = result[previous] + n - 2*count[curr]
            for next in graph[curr]:
                if next == previous:
                    continue
                totalDistance(next, curr)
        if n == 1:
            return [0]
        graph = {}
        for a, b in edges:
            if a not in graph:
                graph[a] = []
            if b not in graph:
                graph[b] = []
            graph[a].append(b)
            graph[b].append(a)
        count = [0] * n
        count_vertices_subtree(0, -1)
        result = [0] * n
        rootDistance(0, -1, 0)
        totalDistance(0, -1)
        return result

File: 834. Sum of Distances in Tree.go
func sumOfDistancesInTree(n int, edges [][]int) []int {
	if n == 1 {
		return []int{0}
	}
	graph := make([][]int, n)
	for _, edge := range edges {
		a, b := edge[0], edge[1]
		graph[a] = append(graph[a], b)
		graph[b] = append(graph[b], a)
	}
	fmt.Println(graph)
	subtreeCount := make([]int, n)
	var getSubtreeCount func(int, int) int
	getSubtreeCount = func(curr, previous int) int {
		count := 1
		for _, next := range graph[curr] {
			if next == previous {
				continue
			}
			count += getSubtreeCount(next, curr)
		}
		subtreeCount[curr] = count
		return count
	}
	getSubtreeCount(0, -1)
	distance := make([]int, n)
	var rootDistance func(int, int, int)
	rootDistance = func(curr, previous, d int) {
		distance[0] += d
		for _, next := range graph[curr] {
			if next == previous {
				continue
			}
			rootDistance(next, curr, d+1)
		}
	}
	rootDistance(0, -1, 0)
	var finalDistance func(int, int)
	finalDistance = func(curr, previous int) {
		if previous != -1 {
			distance[curr] = distance[previous] + n - 2*subtreeCount[curr]
		}
		for _, next := range graph[curr] {
			if next == previous {
				continue
			}
			finalDistance(next, curr)
		}
	}
	finalDistance(0, -1)
	return distance
}